enhancements/workload-partitioning/cpu-limits-workload-partitioning.md

```diff
@@ -0,0 +1,417 @@
+
+As a cluster admin I want to make sure that on a CPU partitioned OpenShift
+cluster, Pods that set CPU limits are also modified with the correct annotation
+for workload partitioning. Furthermore, the CPU limit of those Pods should be
+respected by the container run time so that they are bound to that limit.
+
+### Goals
+
+The goals only apply to clusters where the workload partitioning is enabled.
+
+- Pods that set CPU limits and are annotated for workload partitioning will be
+  modified for workload partitioning.
+- We will not modify the QoS of Pods and guaranteed Pods will not be modified.
+- We will update existing e2e tests to account for this new behavior.
+- Clusters built in this way should pass the same kubernetes and OpenShift
+  conformance and functional end-to-end tests as similar deployments that are
+  not isolating the management workloads.
+
+### Non-Goals
+
+- This enhancement is focused on CPU resources. Other compressible resource
+  types may need to be managed in the future, and those are likely to need
+  different approaches.
+- This enhancement does not address non-compressible resource requests, such as
+  for memory.
+
+## Proposal
+
+In workload partitioning we currently set a custom resource type
+`{workload-type}.workload.openshift.io/cores` for `requests` and `limits`. In
+this process we remove the `requests.cpu` so that we can utilize the existing
+scheduler for assigning pods to nodes. The admission webhook currently skips
+over Pods that set a `limits.cpu` because the default behavior for kubernetes is
+to add `requests.cpu` when `limits.cpu` is set and since we strip the `cpu`
+resource to utilize the scheduler this causes a problem for our desired
+scheduling behavior.
+
+In order to support limits we will need to expose the runtime spec option for
+`CPUQuota` to the workload configuration. There are two components we need to
+modify for this change, CRI-O and the workloads admission webhook.
+
+CRI-O will be updated to expose the CPU quota at the container runtime level via
+the existing workloads configuration. We will modify the existing configuration
+to support the new value. We will make the default behavior to be 0 so existing
+CRI-O configuration files will not need to be modified.
+
+The admission webhook will be altered to no longer ignore modifying CPU limit
+requests. CPU limits will be used and added as annotation to the Pod similar to
+how CPU shares are currently added. The QoS of the Pod shall not be changed, and
+guaranteed Pods will continue to not be altered.
+
+In short CRI-O and the admission webhook will be modified in the following way.
+
+1. CRI-O Workload Resource Configuration
+   - Expose the CPU quota runtime option by adding the `CPUQuota` to the
+     resource configuration.
+   - Update Mutating Spec call to modify CPU Quota
+   - Update Cgroup Manager to set the CPUQuota for workload partitioned
+     containers
+   - Default value of 0 will be assumed for
```

Comment:
should we replace the cpu quota disable annotation with this behavior? if not, how do we reconcile between the two?

Reply:
Good question, as I understand it the cpu quota annotation requires the guaranteed pods, as it stands that's not something we alter so it would seem it's something that can be kept and treated differently. Unless I'm misunderstanding how that annotation gets used.